File: src/AutoContext.Worker.Shared/Logging/CorrelationScope.cs
namespace AutoContext.Worker.Logging;

/// <summary>
/// Worker-process holder for the per-invocation correlation id sent on
/// every <c>TaskRequest</c>. Set by <c>McpToolService</c> for the
/// duration of a task dispatch and read by <c>LogServerLogger</c> at
/// log-emission time so every <c>LogRecord</c> carries the id without
/// requiring callers to thread it through their <c>ILogger</c> APIs.
/// </summary>
/// <remarks>
/// Implemented over <see cref="AsyncLocal{T}"/> so the scope is bound
/// to the async call chain that originated from the dispatch — nested
/// <c>Task</c>s and <c>await</c>s observe the same id, but unrelated
/// dispatches running concurrently each see their own.
/// </remarks>
internal static class CorrelationScope
{
    private static readonly AsyncLocal<string?> Holder = new();

    /// <summary>
    /// The correlation id active on the current async context, or
    /// <c>null</c> when no <see cref="Push"/> scope is in effect.
    /// </summary>
    public static string? Current => Holder.Value;

    /// <summary>
    /// Sets <see cref="Current"/> to <paramref name="correlationId"/>
    /// for the lifetime of the returned <see cref="IDisposable"/>;
    /// disposing restores whatever id was previously in effect.
    /// </summary>
    public static IDisposable Push(string correlationId)
    {
        ArgumentException.ThrowIfNullOrEmpty(correlationId);

        var previous = Holder.Value;
        Holder.Value = correlationId;
        return new Restore(previous);
    }

    private sealed class Restore(string? previous) : IDisposable
    {
        private bool _disposed;

        public void Dispose()
        {
            if (_disposed)
            {
                return;
            }

            _disposed = true;
            Holder.Value = previous;
        }
    }
}

File: src/AutoContext.Worker.Web/src/logging/correlation-scope.ts
import { AsyncLocalStorage } from 'node:async_hooks';

/**
 * Worker-process holder for the per-invocation correlation id sent on
 * every task request envelope. Set by `McpTaskDispatcherService` for the
 * duration of a task dispatch and read by the {@link Logger} at
 * log-emission time so every {@link LogEntry} carries the id without
 * requiring callers to thread it through their logging APIs.
 *
 * Implemented over Node's {@link AsyncLocalStorage} so the scope is
 * bound to the async call chain that originated from the dispatch —
 * nested promises and `await`s observe the same id, but unrelated
 * dispatches running concurrently each see their own.
 *
 * TypeScript counterpart of `CorrelationScope` in
 * `AutoContext.Worker.Shared`.
 */
export class CorrelationScope {
    private static readonly storage = new AsyncLocalStorage<string>();

    /**
     * The correlation id active on the current async context, or
     * `undefined` when no {@link run} scope is in effect.
     */
    static current(): string | undefined {
        return CorrelationScope.storage.getStore();
    }

    /**
     * Runs `callback` with `correlationId` set as the active scope.
     * Resolves with the callback's return value. Restores whatever id
     * (if any) was previously in effect once the callback's promise
     * settles.
     */
    static run<T>(correlationId: string, callback: () => Promise<T>): Promise<T> {
        if (correlationId.length === 0) {
            throw new Error('correlationId must be a non-empty string.');
        }
        return CorrelationScope.storage.run(correlationId, callback);
    }
}
